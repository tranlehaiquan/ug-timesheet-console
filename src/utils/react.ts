import * as React from 'react'

type IdentityFunction = <T>(fn: T) => T

export const typedMemo: IdentityFunction = React.memo

export const getDisplayName = <T extends any = any>(WrappedComponent: React.ComponentType<T>): string => {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
