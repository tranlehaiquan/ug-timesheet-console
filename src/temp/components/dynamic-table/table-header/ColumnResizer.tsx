import * as React from 'react'
import classnames from 'classnames'

interface IResizerProps {
  onResize: React.Dispatch<React.SetStateAction<number|string>>
  min: number
}

export const ColumnResizer: React.FC<IResizerProps> = React.memo(({ onResize, min }) => {

  const ref = React.useRef(null)
  const indicatorRef = React.useRef(null)
  const [isVisible, setIsVisible] = React.useState(false)
  const [isResizing, setIsResizing] = React.useState(false)
  const indicatorHeight = calcInitialIndicatorHeight(ref)

  const onResizeStart = (e: React.MouseEvent) => {
    e.preventDefault()
    setIsResizing(true)
    window.addEventListener('mousemove', onResizeHandler)
    window.addEventListener('mouseup', onResizeEnd)
  }

  const onResizeEnd = (e: MouseEvent) => {
    e.preventDefault()
    setIsResizing(false)
    setIsVisible(false)
    window.removeEventListener('mousemove', onResizeHandler)
    window.removeEventListener('mouseup', onResizeEnd)
  }

  const onResizeHandler = (e: MouseEvent) => {
    e.preventDefault()
    const thElement = ref.current.parentElement
    const currentWidth = thElement.offsetWidth
    const newWidth = Math.max(min, currentWidth + e.movementX)

    if (newWidth !== currentWidth) {
      const tab = ref.current.closest('table')
      tab.style.width = `${tab.offsetWidth + e.movementX}px`
      onResize(newWidth)
      indicatorRef.current.style.height = `${tab.clientHeight - 1}px`
    }
  }

  const onShowHandler = (e: React.MouseEvent) => {
    e.preventDefault()
    setIsVisible(true)
  }

  const onHideHandler = (e: React.MouseEvent) => {
    e.preventDefault()
    if (!isResizing) {
      setIsVisible(false)
    }
  }

  const visibilityClass = isVisible ? 'sk-visible' : 'sk-hidden'

  return (
    <span
      data-sk-name="sk-resize-container"
      ref={ ref }
      onMouseDown={ onResizeStart }
      onMouseEnter={ onShowHandler }
      onMouseLeave={ onHideHandler }
      className="sked-column-resizer-container"
    >
      <span
        data-sk-name="sk-resize-indicator"
        ref={ indicatorRef }
        className={ classnames('sked-column-resizer-indicator', visibilityClass) }
        style={ { height: indicatorHeight } }
      />
      <span
        data-sk-name="sk-resize-handle"
        className={ classnames('sked-column-resizer-handle', visibilityClass) }
      />
    </span>
  )
})

const calcInitialIndicatorHeight = (containerRef: React.RefObject<HTMLSpanElement>) => {
  return containerRef.current && containerRef.current.closest('table')
    ? containerRef.current.closest('table').clientHeight - 1
    : 0
}
