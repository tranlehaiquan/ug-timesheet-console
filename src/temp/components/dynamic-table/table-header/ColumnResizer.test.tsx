import * as React from 'react'
import { shallow, mount } from 'enzyme'

import { ColumnResizer } from './ColumnResizer'

const selectors = {
  container: '[data-sk-name="sk-resize-container"]',
  indicator: '[data-sk-name="sk-resize-indicator"]',
  handle: '[data-sk-name="sk-resize-handle"]'
}

const sampleOnResizeHandler = (width: number) => { }

const getTableWithResizer = () => {
  return (
    <table>
      <thead>
        <tr>
          <th>
            <p>sample_cell_content</p>
            <ColumnResizer
              onResize={ sampleOnResizeHandler }
              min={ 120 }
            />
          </th>
        </tr>
      </thead>
    </table>
  )
}

describe('ColumnResizer', () => {

  test('Renders properly', () => {
    const table = mount(getTableWithResizer())
    expect(table).toMatchSnapshot()
    expect(table.find(selectors.container).length).toEqual(1)
    expect(table.find(selectors.indicator).length).toEqual(1)
    expect(table.find(selectors.handle).length).toEqual(1)
  })

  test('Is hidden by default', () => {
    const table = mount(getTableWithResizer())
    const indicator = table.find(selectors.indicator)
    expect(indicator.hasClass('sk-visible')).toEqual(false)
    expect(indicator.hasClass('sk-hidden')).toEqual(true)
    const handle = table.find(selectors.handle)
    expect(handle.hasClass('sk-visible')).toEqual(false)
    expect(handle.hasClass('sk-hidden')).toEqual(true)
  })

  test('Shows handle and indicator on mouseenter event', () => {
    const table = mount(getTableWithResizer())
    const container = table.find(selectors.container)

    let indicator = table.find(selectors.indicator)
    let handle = table.find(selectors.handle)
    expect(indicator.hasClass('sk-visible')).toEqual(false)
    expect(indicator.hasClass('sk-hidden')).toEqual(true)
    expect(handle.hasClass('sk-visible')).toEqual(false)
    expect(handle.hasClass('sk-hidden')).toEqual(true)

    container.simulate('mouseenter')
    indicator = table.find(selectors.indicator)
    handle = table.find(selectors.handle)
    expect(indicator.hasClass('sk-visible')).toEqual(true)
    expect(indicator.hasClass('sk-hidden')).toEqual(false)
    expect(handle.hasClass('sk-visible')).toEqual(true)
    expect(handle.hasClass('sk-hidden')).toEqual(false)
  })

  test('Hides handle and indicator on mouseleave event', () => {
    const table = mount(getTableWithResizer())
    const container = table.find(selectors.container)

    let indicator = table.find(selectors.indicator)
    let handle = table.find(selectors.handle)
    expect(indicator.hasClass('sk-visible')).toEqual(false)
    expect(indicator.hasClass('sk-hidden')).toEqual(true)
    expect(handle.hasClass('sk-visible')).toEqual(false)
    expect(handle.hasClass('sk-hidden')).toEqual(true)

    container.simulate('mouseenter')
    container.simulate('mouseleave')

    indicator = table.find(selectors.indicator)
    handle = table.find(selectors.handle)
    expect(indicator.hasClass('sk-visible')).toEqual(false)
    expect(indicator.hasClass('sk-hidden')).toEqual(true)
    expect(handle.hasClass('sk-visible')).toEqual(false)
    expect(handle.hasClass('sk-hidden')).toEqual(true)
  })

})
