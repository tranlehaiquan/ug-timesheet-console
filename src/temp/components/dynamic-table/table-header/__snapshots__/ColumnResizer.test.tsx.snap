// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColumnResizer Renders properly 1`] = `
<table>
  <thead>
    <tr>
      <th>
        <p>
          sample_cell_content
        </p>
        <Component
          min={120}
          onResize={[Function]}
        >
          <span
            className="sked-column-resizer-container"
            data-sk-name="sk-resize-container"
            onMouseDown={[Function]}
            onMouseEnter={[Function]}
            onMouseLeave={[Function]}
          >
            <span
              className="sked-column-resizer-indicator sk-hidden"
              data-sk-name="sk-resize-indicator"
              style={
                Object {
                  "height": 0,
                }
              }
            />
            <span
              className="sked-column-resizer-handle sk-hidden"
              data-sk-name="sk-resize-handle"
            />
          </span>
        </Component>
      </th>
    </tr>
  </thead>
</table>
`;
